### PR TITLE
Implement RFC 2523, `#[cfg(version(..))]`

### PR DESCRIPTION
**Pull request by mibac138**
Hi! This is my first contribution to rust, I hope I didn't miss anything. I tried to implement this feature so that `#[cfg(version(1.44.0))]` works but the parser was printing an error that I wasn't sure how to fix so I just opted for implementing `#[cfg(version("1.44.0"))]` (note the quotes).

Tracking issue: #64796

### CODE REVIEW

File: src/librustc_attr/builtin.rs [hunk 1]
```diff
@@ -564,11 +565,8 @@ pub fn find_crate_name(attrs: &[Attribute]) -> Option<Symbol> {
 
 /// Tests if a cfg-pattern matches the cfg set
 pub fn cfg_matches(cfg: &ast::MetaItem, sess: &ParseSess, features: Option<&Features>) -> bool {
-    eval_condition(cfg, sess, &mut |cfg| {
-        let gate = find_gated_cfg(|sym| cfg.check_name(sym));
-        if let (Some(feats), Some(gated_cfg)) = (features, gate) {
-            gate_cfg(&gated_cfg, cfg.span, sess, feats);
-        }
+    eval_condition(cfg, sess, features, &mut |cfg| {
+        try_gate_cfg(cfg, sess, features);
```
Comment (pickfire) [inline src/librustc_attr/builtin.rs:569]: I wonder should `try_gate_cfg` be moved into `eval_condition`? Is this the right place?

File: src/doc/unstable-book/src/language-features/cfg-version.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+depending on the compiler version.
+
+## Examples
+
+```rust
+#![feature(cfg_version)]
+
+#[cfg(version("1.42"))]
+fn a() {
+  // ...
```
Comment (pickfire) [inline src/doc/unstable-book/src/language-features/cfg-version.md:19]: Should the examples here be 4 spaces?

File: src/librustc_attr/builtin.rs [hunk 3]
```diff
@@ -612,9 +617,42 @@ fn gate_cfg(gated_cfg: &GatedCfg, cfg_span: Span, sess: &ParseSess, features: &F
...
+                Some(ver) => ver,
+                None => {
+                    sess.span_diagnostic.struct_span_err(*span, "invalid version string").emit();
+                    return false;
+                }
+            };
+            let version = option_env!("CFG_VERSION").unwrap_or("unknown rustc version");
+            let version = Version::parse(version).unwrap();
+
+            return version >= min_version;
```
Comment (pickfire) [inline src/librustc_attr/builtin.rs:654]:
```suggestion
            version >= min_version
```

File: src/test/ui/feature-gates/feature-gate-cfg-version.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+fn bar() -> bool  { false }
+#[cfg(version("foo"))] //~ ERROR: invalid version string
+//~^ ERROR `cfg(version)` is experimental and subject to change
+fn bar() -> bool  { false }
+#[cfg(version("999"))]
+//~^ ERROR `cfg(version)` is experimental and subject to change
+fn bar() -> bool  { false }
+#[cfg(version("0"))]
+//~^ ERROR `cfg(version)` is experimental and subject to change
+fn bar() -> bool { true }
```
Comment (pickfire) [inline src/test/ui/feature-gates/feature-gate-cfg-version.rs:22]: Should we add a test for `#[cfg(not(version("1.42")))]` too?

**Comment by pickfire** [comment]
As per the RFC, should not it be `#[cfg(version(1.27))]` rather than `#[cfg(version("1.27"))]`?

File: src/librustc_attr/builtin.rs (hunk 1 above)
Comment (mibac138) [reply src/librustc_attr/builtin.rs:569]: I think this should work. I'm testing locally right now

File: src/doc/unstable-book/src/language-features/cfg-version.md (hunk 2 above)
Comment (mibac138) [reply src/doc/unstable-book/src/language-features/cfg-version.md:19]: Oops, I just mostly copy-pasted from `cfg-sanitize.md` and didn't even notice. Thanks!

File: src/librustc_attr/builtin.rs (hunk 3 above)
Comment (mibac138) [reply src/librustc_attr/builtin.rs:654]: Thanks, nice catch

File: src/test/ui/feature-gates/feature-gate-cfg-version.rs (hunk 4 above)
Comment (mibac138) [reply src/test/ui/feature-gates/feature-gate-cfg-version.rs:22]: @pickfire I believe I already added a test like this at lines 4 to 6 (`1.44` is the current nightly version)?

**Comment by mibac138** [comment]
@pickfire I did try to implement it that way but the parser was emitting an error (`expected `,`, found `.`` whenever I specified a full verison, so `1.42.0` errors, `1.42` doesn't) and I don't know what can I do prevent that.

**Comment by pickfire** [comment]
I think that is because it is not a valid integer to be parsed. It could be a float with 1.27 but not 1.27.0 which means nothing. Maybe we should ask around, I am not quite sure on the rust internals.

File: src/doc/unstable-book/src/language-features/cfg-sanitize.md [hunk 5]
```diff
@@ -11,26 +11,24 @@ depending on whether a particular sanitizer is enabled or not.
 
 ## Examples
```
Comment (pickfire) [inline src/doc/unstable-book/src/language-features/cfg-sanitize.md:12]: I think sanitize should not be in this pull request, it should be another pull request.

Comment (mibac138) [reply src/doc/unstable-book/src/language-features/cfg-sanitize.md:12]: I don't think this is a significant enough change to warrant it's own PR. However, if `@petrochenkov` or `@varkor` disagree I'll split it into 2 PRs.

**Comment by pickfire** [comment]
@mibac138 I believe accepting `1.27.0` requires changing the parser.

File: src/doc/unstable-book/src/language-features/cfg-sanitize.md (hunk 5 above)
Comment (petrochenkov) [reply src/doc/unstable-book/src/language-features/cfg-sanitize.md:12]: Yeah, if the change is entirely unrelated to the main theme of the PR, it's better to make a separate PR for it. Pull requests can be made for arbitrarily small changes.

File: src/librustc_parse/parser/attr.rs [hunk 6]
```diff
@@ -273,17 +273,27 @@ impl<'a> Parser<'a> {
...
     }
 
-    crate fn parse_meta_item_kind(&mut self) -> PResult<'a, ast::MetaItemKind> {
+    crate fn parse_meta_item_kind(&mut self, path: &ast::Path) -> PResult<'a, ast::MetaItemKind> {
         Ok(if self.eat(&token::Eq) {
             ast::MetaItemKind::NameValue(self.parse_unsuffixed_lit()?)
         } else if self.check(&token::OpenDelim(token::Paren)) {
-            // Matches `meta_seq = ( COMMASEP(meta_item_inner) )`.
-            let (list, _) = self.parse_paren_comma_seq(|p| p.parse_meta_item_inner())?;
+            let (list, _) = if path == &rustc_span::sym::version {
```
Comment (petrochenkov) [inline src/librustc_parse/parser/attr.rs:285]: This is incorrect, `version(...)` shouldn't allow semver expression like `1.2.3` in arbitrary attributes, only in `cfg`.

Comment (petrochenkov) [reply src/librustc_parse/parser/attr.rs:285]: As a result, `MetaItemKind::List` cannot be used for `version(1.2.3)` in `eval_condition` because it's not a meta-item.

You also cannot use `span_to_snippet` for obtaining the text representation of `1.2.3` in general because spans are not reliable.

I'd recommend implementing `version("1.2.3")` with a string literal in an initial PR (which *is* a meta-item and is already stringified), and then think about parsing and stringifying `1.2.3` instead.

Comment (pickfire) [reply src/librustc_parse/parser/attr.rs:285]: "1.2.3" was already done in the previous commits.

Comment (pickfire) [reply src/librustc_parse/parser/attr.rs:285]: The RFC uses `version(1.27.0)` as example IIRC.

File: src/librustc_attr/builtin.rs [hunk 7]
```diff
@@ -623,28 +623,22 @@ pub fn eval_condition(
...
-                        .struct_span_err(*span, &*format!("expected string literal"))
-                        .emit();
-                    return false;
-                }
-                [..] => {
+            let span = match &mis[..] {
+                [] => {
                     sess.span_diagnostic
-                        .struct_span_err(cfg.span, "expected single string literal")
+                        .struct_span_err(cfg.span, "expected a version string")
```
Comment (pickfire) [inline src/librustc_attr/builtin.rs:629]: I believe it may be better to use `expected a version literal`.

File: src/test/ui/feature-gates/feature-gate-cfg-version.rs [hunk 8]
```diff
@@ -1,28 +1,24 @@
...
+#[cfg(version(1.45))]
 //~^ ERROR `cfg(version)` is experimental and subject to change
 fn foo() -> bool { true }
-#[cfg(not(version("1.44")))]
+#[cfg(not(version(1.45)))]
 //~^ ERROR `cfg(version)` is experimental and subject to change
 fn foo() -> bool { false }
 
-#[cfg(version("1.43", "1.44", "1.45"))] //~ ERROR: expected single string literal
+#[cfg(version(99.99.99))]
```
Comment (pickfire) [inline src/test/ui/feature-gates/feature-gate-cfg-version.rs:8]: Is there a test for an overflowing version? Such as `99999`?

Comment (pickfire) [reply src/test/ui/feature-gates/feature-gate-cfg-version.rs:8]: Or how about negative version?

File: src/doc/unstable-book/src/language-features/cfg-sanitize.md (hunk 5 above)
Comment (mibac138) [reply src/doc/unstable-book/src/language-features/cfg-sanitize.md:12]: Opened #71637

File: src/librustc_attr/builtin.rs (hunk 7 above)
Comment (mibac138) [reply src/librustc_attr/builtin.rs:629]: It does seem better, changed it.

File: src/test/ui/feature-gates/feature-gate-cfg-version.rs (hunk 8 above)
Comment (mibac138) [reply src/test/ui/feature-gates/feature-gate-cfg-version.rs:8]: Good idea, added overflow and underflow tests

File: src/librustc_parse/parser/attr.rs (hunk 6 above)
Comment (mibac138) [reply src/librustc_parse/parser/attr.rs:285]: Ok, I'm going to go back to the original string literal version

File: src/librustc_attr/builtin.rs [hunk 9]
```diff
@@ -612,9 +617,42 @@ fn gate_cfg(gated_cfg: &GatedCfg, cfg_span: Span, sess: &ParseSess, features: &F
...
+                }
+            };
+            let min_version = match Version::parse(&min_version.as_str()) {
+                Some(ver) => ver,
+                None => {
+                    sess.span_diagnostic.struct_span_err(*span, "invalid version literal").emit();
+                    return false;
+                }
+            };
+            let version = option_env!("CFG_VERSION").unwrap_or("unknown rustc version");
```
Comment (pickfire) [inline src/librustc_attr/builtin.rs:651]: Should we use?
```suggestion
            let version = env!("CFG_VERSION");
```
I don't think the `unknown rustc version` is a good error message here, the default for `env!` looks well enough.

File: src/test/ui/feature-gates/feature-gate-cfg-version.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+fn bar() -> bool  { false }
+#[cfg(version("65536"))] //~ ERROR: invalid version literal
+//~^ ERROR `cfg(version)` is experimental and subject to change
+fn bar() -> bool  { false }
+#[cfg(version("0"))]
+//~^ ERROR `cfg(version)` is experimental and subject to change
+fn bar() -> bool { true }
+
+#[cfg(version("1.65536.2"))]
+fn version_check_bug() {}
```
Comment (pickfire) [inline src/test/ui/feature-gates/feature-gate-cfg-version.rs:31]: Wow, do we need this? Or maybe we can try another crate `semver`?

Comment (mibac138) [reply src/test/ui/feature-gates/feature-gate-cfg-version.rs:31]: I think this it's unlikely for anyone to run into this, but I wanted to explicitly document this problem. I have thought about using `semver`, however if I were to use `semver::Version` I'd need to make sure the version contains major, minor and patch versions, and if I were to use `semver::VersionReq` I'd need to check if the version literal doesn't contain things like `>=1.0`, `<1.5`, `>1.2, <1.3` which the RFC only mentions as possible future work ([`However, as a first iteration, version(1.27.0) is simple and covers most use cases.`](https://github.com/rust-lang/rfcs/blob/master/text/2523-cfg-path-version.md#the-bikeshed---attribute-name)). Unless it's okay to implement this using `VersionReq` and adding support for `>`, `<`, etc. syntax I think it's simpler to just use `version_check` despite this bug which I believe is unlikely to affect anyone.

File: src/librustc_attr/builtin.rs (hunk 9 above)
Comment (mibac138) [reply src/librustc_attr/builtin.rs:651]: Now that I think about it what I meant to do is `env!("CFG_VERSION", "unknown rustc version")` but I agree that the default `env` error message seems better

Comment (pickfire) [reply src/librustc_attr/builtin.rs]: I tried `env!("CFG_VERSION", "unknown rustc version")` but the result isn't good.

**Review by pickfire — APPROVED** [review]
Looks good to me.

File: src/librustc_attr/builtin.rs [hunk 11]
```diff
@@ -612,9 +617,41 @@ fn gate_cfg(gated_cfg: &GatedCfg, cfg_span: Span, sess: &ParseSess, features: &F
...
+                }
+            };
+            let min_version = match Version::parse(&min_version.as_str()) {
+                Some(ver) => ver,
+                None => {
+                    sess.span_diagnostic.struct_span_err(*span, "invalid version literal").emit();
+                    return false;
+                }
+            };
+            let version = Version::parse(env!("CFG_VERSION")).unwrap();
```
Comment (petrochenkov) [inline src/librustc_attr/builtin.rs:651]: Due to this use of an environment variable, the `rustc_attr` crate needs a build script now.
```rust
fn main() {
    println!("cargo:rerun-if-changed=build.rs");
    println!("cargo:rerun-if-env-changed=CFG_VERSION");
}
```

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by mibac138** [comment]
Thank you @petrochenkov and @pickfire for your reviews and help! I found them helpful :)

I have rebased my PR and resolved the merge conflict.
